File: src/main/java/network/packet/incoming/impl/DialoguePacketReader.java
package network.packet.incoming.impl;

import game.entity.actor.player.Player;
import network.packet.Packet;
import network.packet.context.PacketContext;
import network.packet.context.impl.DialogueChatPacketContext;
import network.packet.incoming.IncomingPacketReader;
import utility.functions.Misc;
import utility.game.InputEvent;

/**
 * @author Tyluur <itstyluur@icloud.com>
 * @since 2019-02-04
 */
public class DialoguePacketReader implements IncomingPacketReader {
	
	@Override
	public int[] bindings() {
		return arguments(DIALOGUE_CONTINUE_PACKET, ENTER_STRING_PACKET, ENTER_LONG_STRING_PACKET, ENTER_INTEGER_PACKET);
	}
	
	@Override
	public PacketContext read(Player player, Packet stream) {
		switch (stream.getOpcode()) {
			case DIALOGUE_CONTINUE_PACKET: {
				int interfaceHash = stream.readIntV2();
				@SuppressWarnings("unused") int junk = stream.readShortLE128();
				int interfaceId = interfaceHash >> 16;
				@SuppressWarnings("unused") int buttonId = (interfaceHash & 0xFF);
				if (Misc.getInterfaceDefinitionsSize() <= interfaceId) {
					break;
				}
				if (!player.isRunning() || !player.getInterfaceManager().containsInterface(interfaceId)) {
					break;
				}
				int componentId = interfaceHash - (interfaceId << 16);
				return new DialogueChatPacketContext(interfaceId, componentId);
			}
			case ENTER_STRING_PACKET: {
				if (!player.isRunning() || player.isDead()) {
					break;
				}
				String value = stream.readRS2String();
				if (value.equals("")) {
					break;
				}
				return new PacketContext() {
					@Override
					public void handle(Player player) {
						if (player.getTemporaryAttribute("input_event", null) != null) {
							InputEvent event = player.removeTemporaryAttribute("input_event");
							event.setInput(value);
							event.handleInput();
							return;
						}
						if (player.getInterfaceManager().containsInterface(1108)) {
							player.getContactManager().setChatPrefix(value);
						}
					}
				};
			}
			case ENTER_LONG_STRING_PACKET: {
				String value = stream.readRS2String();
				if (value.equals("")) {
					break;
				}
				return new PacketContext() {
					@Override
					public void handle(Player player) {
						if (player.getTemporaryAttribute("input_event") != null) {
							InputEvent event = player.removeTemporaryAttribute("input_event");
							event.setInput(value);
							event.handleInput();
						}
					}
				};
			}
			case ENTER_INTEGER_PACKET: {
				if (!player.isRunning() || player.isDead()) {
					break;
				}
				int value = stream.readInt();
				return new PacketContext() {
					@Override
					public void handle(Player player) {
						if (player.getTemporaryAttribute("input_event") != null) {
							InputEvent event = player.removeTemporaryAttribute("input_event");
							event.setInput(value);
							event.handleInput();
						}
					}
				};
			}
		}
		return null;
	}
}

File: src/main/java/org/redrune/networking/packet/incoming/impl/DialoguePacketReader.kt
package org.redrune.networking.packet.incoming.impl

import org.redrune.game.entity.actor.player.Player
import org.redrune.networking.packet.Packet
import org.redrune.networking.packet.context.PacketContext
import org.redrune.networking.packet.context.impl.DialogueChatPacketContext
import org.redrune.networking.packet.incoming.IncomingPacketReader
import org.redrune.utility.constants.PacketConstants
import org.redrune.utility.functions.Misc
import org.redrune.utility.game.InputEvent

/**
 * @author Tyluur <itstyluur@icloud.com>
 * @since 2019-02-04
 */
class DialoguePacketReader : IncomingPacketReader {

    override fun bindings(): IntArray {
        return arguments(
            PacketConstants.DIALOGUE_CONTINUE_PACKET,
            PacketConstants.ENTER_STRING_PACKET,
            PacketConstants.ENTER_LONG_STRING_PACKET,
            PacketConstants.ENTER_INTEGER_PACKET
        )
    }

    override fun read(player: Player, stream: Packet): PacketContext? {
        when (stream.opcode) {
            PacketConstants.DIALOGUE_CONTINUE_PACKET -> {
                val interfaceHash = stream.readIntV2()
                @Suppress("unused") val junk = stream.readShortLE128()
                val interfaceId = interfaceHash shr 16
                @Suppress("unused") val buttonId = interfaceHash and 0xFF
                if (Misc.getInterfaceDefinitionsSize() <= interfaceId) {
                    return null
                }
                if (!player.isRunning || !player.interfaceManager.containsInterface(interfaceId)) {
                    return null
                }
                val componentId = interfaceHash - (interfaceId shl 16)
                return DialogueChatPacketContext(interfaceId, componentId)
            }

            PacketConstants.ENTER_STRING_PACKET -> {
                if (!player.isRunning || player.isDead) {
                    return null
                }
                val value = stream.readRS2String()
                if (value == "") {
                    return null
                }
                return object : PacketContext() {
                    override fun handle(player: Player) {
                        if (player.getTemporaryAttribute<Any?>("input_event", null) != null) {
                            val event = player.removeTemporaryAttribute<InputEvent>("input_event")
                            event.setInput(value)
                            event.handleInput()
                            return
                        }
                        if (player.interfaceManager.containsInterface(1108)) {
                            player.contactManager.setChatPrefix(value)
                        }
                    }
                }
            }

            PacketConstants.ENTER_LONG_STRING_PACKET -> {
                val value = stream.readRS2String()
                if (value == "") {
                    return null
                }
                return object : PacketContext() {
                    override fun handle(player: Player) {
                        if (player.getTemporaryAttribute<Any?>("input_event") != null) {
                            val event = player.removeTemporaryAttribute<InputEvent>("input_event")
                            event.setInput(value)
                            event.handleInput()
                        }
                    }
                }
            }

            PacketConstants.ENTER_INTEGER_PACKET -> {
                if (!player.isRunning || player.isDead) {
                    return null
                }
                val value = stream.readInt()
                return object : PacketContext() {
                    override fun handle(player: Player) {
                        if (player.getTemporaryAttribute<Any?>("input_event") != null) {
                            val event = player.removeTemporaryAttribute<InputEvent>("input_event")
                            event.setInput(value)
                            event.handleInput()
                        }
                    }
                }
            }
        }
        return null
    }
}